couchdb/client.py

```diff
@@ -775,10 +775,161 @@ def put_attachment(self, doc, content, filename=None, content_type=None):
+        ...          'fields': ['name'],
+        ...          'sort':[{'name': 'asc'}]}
+        >>> for row in db.find(mango):                          # doctest: +SKIP
+        ...    print(row['name'])                               # doctest: +SKIP
+        John Doe
+        Mary Jane
+        >>> del server['python-tests']
+
+        :param mango_query: a dictionary describing criteria used to select
+                            documents
+        :param wrapper: an optional callable that should be used to wrap the
+                        resulting documents
+        :return: the query results as a list of `Document` (or whatever `wrapper` returns)
+        """
+        assert isinstance(mango_query, dict)
+        status, headers, data = self.resource.post_json('_find', mango_query)
+        wrapper = wrapper or Document
+        return map(wrapper, data.get('docs', []))
+
+    def explain(self, mango_query):
+        """Explain a mango find-query.
+
+        Note: Only available for CouchDB version >= 2.0.0 .
+
+        More information on the `mango_query` structure can be found here:
+          http://docs.couchdb.org/en/master/api/database/find.html#db-explain
+
+        >>> server = Server()
+        >>> db = server.create('python-tests')
+        >>> db['johndoe'] = dict(type='Person', name='John Doe')
+        >>> db['maryjane'] = dict(type='Person', name='Mary Jane')
+        >>> db['gotham'] = dict(type='City', name='Gotham City')
+        >>> mango = {'selector': {'type': 'Person'}, 'fields': ['name']}
+        >>> db.explain(mango)                          #doctest: +ELLIPSIS +SKIP
+        {...}
+        >>> del server['python-tests']
+
+        :param mango_query: a `dict` describing criteria used to select
+                            documents
+        :return: the query results as a list of `Document` (or whatever
+                 `wrapper` returns)
+        :rtype: `dict`
+        """
+        assert isinstance(mango_query, dict)
+        _, _, data = self.resource.post_json('_explain', mango_query)
+        return data
+
+    def index(self):
+        """Get all available indexes.
+
+        Note: Only available for CouchDB version >= 2.0.0 .
+
+        More information here:
+          http://docs.couchdb.org/en/master/api/database/find.html#get--db-_index
+
+        >>> server = Server()
+        >>> db = server.create('python-tests')
+        >>> db.index()                                           #doctest: +SKIP
+        {'indexes': [{'ddoc': None,
+          'def': {'fields': [{'_id': 'asc'}]},
+          'name': '_all_docs',
+          'type': 'special'}],
+        'total_rows': 1}
+        >>> del server['python-tests']
+
+        :return: `dict` containing the number of indexes (`total_rows`) and
+                 a description of each index (`indexes`)
+        """
+        _, _, data = self.resource.get_json('_index')
+        return data
+
+    def add_index(self, index, ddoc=None, name=None):
```

Comment:
Instead of having separate `index()`, `add_index()` and `remove_index()` methods, did you give any thought to somehow unifying the API, in a way that both unclutters the API a bit as well as better mirrors CouchDB's HTTP API? Maybe there could be an `index()` method that returns an `Indexes` object, which could then have `__setitem__` and `__delitem__` methods to deal with adding (as well as replacing, I guess) and removing?